### PR TITLE
alsa-firmware: enable cross compilation

### PR DESCRIPTION
**Pull request by dingxiangfei2009**
###### Motivation for this change

The Autoconf and Automake scripts in `alsa-firmware` incorrectly build the firmware writers to the host platform, which leads to incorrect Exec format if they are invoked by these scripts later to transform the firmware blobs. This PR will properly build the writers to the build platform instead.
###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nix-review --run "nix-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Assured whether relevant documentation is up to date
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

### CODE REVIEW

File: pkgs/os-specific/linux/alsa-firmware/cross.patch [hunk 1]
```diff
@@ -0,0 +1,347 @@
+--- a/hdsploader/Makefile.am	2015-02-26 20:36:03.000000000 +0800
```
Comment (JohnAZoidberg) [inline pkgs/os-specific/linux/alsa-firmware/cross.patch]: Where did you get this patch from?

File: pkgs/os-specific/linux/alsa-firmware/default.nix [hunk 2]
```diff
@@ -8,6 +8,19 @@ stdenv.mkDerivation rec {
     sha256 = "0gfcyj5anckjn030wcxx5v2xk2s219nyf99s9m833275b5wz2piw";
   };
 
+  patches = [ ./cross.patch ];
+
+  nativeBuildInputs = [ automake autoconf buildPackages.stdenv.cc ];
```
Comment (JohnAZoidberg) [inline pkgs/os-specific/linux/alsa-firmware/default.nix]:
```suggestion
  nativeBuildInputs = [ autoreconfHook buildPackages.stdenv.cc ];
```

 This hook does all the autotools steps, so you don't need the `preConfigure` either.

File: pkgs/os-specific/linux/alsa-firmware/default.nix [hunk 3]
```diff
@@ -1,4 +1,4 @@
-{stdenv, fetchurl}:
+{stdenv, buildPackages, autoconf, automake, fetchurl}:
```
Comment (JohnAZoidberg) [inline pkgs/os-specific/linux/alsa-firmware/default.nix]:
```suggestion
{ stdenv, buildPackages, autoreconfHook, fetchurl }:
```

Spaces + autoreconfHook

**Review by JohnAZoidberg — CHANGES_REQUESTED** [review]
For my host system it builds fine but cross compiling fails:
```
builder for '/nix/store/gb1blacz8nlm8srjznpl7njspq74g51z-alsa-firmware-1.0.29-aarch64-unknown-linux-gnu.drv' failed with exit code 2; last 10 log lines:
  make[1]: Leaving directory '/build/alsa-firmware-1.0.29/korg1212'
  Making all in maestro3
  make[1]: Entering directory '/build/alsa-firmware-1.0.29/maestro3'
  aarch64-unknown-linux-gnu-gcc -DPACKAGE_NAME=\"alsa-firmware\" -DPACKAGE_TARNAME=\"alsa-firmware\" -DPACKAGE_VERSION=\"1.0.29\" -DPACKAGE_STRING=\"alsa-firmware\ 1.0.29\" -DPACKAGE_BUGREPORT=\"\" -DPACKAGE_URL=\"\" -DSTDC_HEADERS=1 -DPAC
KAGE=\"alsa-firmware\" -DVERSION=\"1.0.29\" -I.     -g -O2 -c -o fw_writer.o fw_writer.c
  aarch64-unknown-linux-gnu-gcc  -g -O2   -o fw_writer fw_writer.o  
  ./fw_writer
  /nix/store/x6b81sfmbmkmcyqp2rmk6jgsmzvyrjj0-bash-4.4-p23/bin/bash: ./fw_writer: cannot execute binary file: Exec format error
  make[1]: *** [Makefile:612: maestro3_assp_kernel.fw] Error 126
  make[1]: Leaving directory '/build/alsa-firmware-1.0.29/maestro3'
  make: *** [Makefile:361: all-recursive] Error 1
[0 built (1 failed)]
error: build of '/nix/store/gb1blacz8nlm8srjznpl7njspq74g51z-alsa-firmware-1.0.29-aarch64-unknown-linux-gnu.drv' failed
```

Looks like you need to do the same for `fw_writer` as for `tobin`.

**Comment by JohnAZoidberg** [comment]
It seems like it builds the `tobin` program and runs that during the build process.
See line 52 in `hdsploader/Makefile.in`: `noinst_PROGRAMS = tobin$(EXEEXT)`
Not sure how to build that for the build architecture and the rest for the target.

**Comment by JohnAZoidberg** [comment]
You need to do the same to `fw_writer` in `{maestro3,sb16,wavefront,ymfpci}/Makefile.am`
```diff
-$(firmware_files): fw_writer
-	./fw_writer
+LINK_FOR_BUILD.c = $(CC_FOR_BUILD) $(CFLAGS_FOR_BUILD) $(CPPFLAGS_FOR_BUILD) $(LDFLAGS_FOR_BUILD) $(TARGET_ARCH_FOR_BUILD)
+
+$(fw_writer_OBJECTS) : CC=$(CC_FOR_BUILD)
+$(fw_writer_OBJECTS) : CFLAGS=$(CFLAGS_FOR_BUILD)
+$(fw_writer_OBJECTS) : CPPFLAGS=$(CPPFLAGS_FOR_BUILD)
+
+fw_writer$(BUILD_EXEEXT): $(tobin_OBJECTS)
+	$(LINK_FOR_BUILD.c) $^ $(LOADLIBES_FOR_BUILD) $(LDLIBS_FOR_BUILD) -o $@
+
+$(firmware_files): fw_writer$(BUILD_EXEEXT)
+	./fw_writer$(BUILD_EXEEXT)```

Then it builds fine for me :)

File: pkgs/os-specific/linux/alsa-firmware/cross.patch (hunk 1 above)
Comment (dingxiangfei2009) [reply pkgs/os-specific/linux/alsa-firmware/cross.patch]: I composed this patch myself. :sweat_smile:

File: pkgs/os-specific/linux/alsa-firmware/default.nix (hunk 2 above)
Comment (dingxiangfei2009) [reply pkgs/os-specific/linux/alsa-firmware/default.nix]: That is neat!

**Comment by dingxiangfei2009** [comment]
@JohnAZoidberg I don't know. I built `pkgsCross.raspberryPi.alsa-firmware` and `maestro3` did not complain. Also, `maestro3/Makefile.am` is indeed in the patch.

**Comment by JohnAZoidberg** [comment]
Thanks, now your command for building for raspi and `nix build -f . alsa-firmware --arg crossSystem '{ config = "aarch64-unknown-linux-gnu"; }'` works.

**Comment by JohnAZoidberg** [comment]
Have you tried getting the patch merged upstream? :)

**Review by JohnAZoidberg — APPROVED** [review]

**Comment by dingxiangfei2009** [comment]
@JohnAZoidberg Oh, good point. I should try upstreaming it, too.

**Comment by lopsided98** [comment]
I see that you tried to [upstream this patch](https://www.spinics.net/lists/alsa-devel/msg93454.html), but it looks like it was ignored. Maybe you should replace the patch with a `fetchpatch` invocation that pulls from patchwork (https://patchwork.kernel.org/patch/11028517/mbox/), if only to document that an attempt was made to upstream it?

In any case, I tested cross builds for armv6l, armv7l and aarch64 and they all succeeded. I would love to get this merged, as it is one of the few blockers for basic cross-compiled NixOS.

**Comment by dingxiangfei2009** [comment]
@lopsided98 sad to hear. Okay, I will use a `fetchpatch` then.

**Comment by erictapen** [comment]
This is also working fine for me on `aarch64`. Would love to see this merged soon!

**Comment by dingxiangfei2009** [comment]
@JohnAZoidberg @matthewbauer Okay, I think I need to update this PR to the latest master (?) to keep it fresh, but should I rebase or merge?

**Comment by dingxiangfei2009** [comment]
@lopsided98 Maybe it was my fault. I guess the email client I used to submit this patch did something funny with the patch file, and it looks corrupted. I will try to rectify this upstream patch.

**Comment by JohnAZoidberg** [comment]
What's the status of this @dingxiangfei2009?
Did you fix the patch?

Please rebase it on top of the lastest master.

**Comment by Infinisil** [comment]
Any interest in this still?

**Comment by Atemu** [comment]
I just ran into this issue while trying to cross-compile a very basic sd-card image for armv7l-hf-multiplatform and this patch solved it

**Comment by matthewbauer** [comment]
It looks like this no longer applies as of alsa-1.2.1.

**Comment by matthewbauer** [comment]
Updated version at https://github.com/matthewbauer/alsa-firmware/commit/2b771ecbff645d720e9766279987e83ec987c7ee

**Comment by erictapen** [comment]
Thanks for merging, @matthewbauer. Any chance this could get backported to `nixos-20.03`?

**Pull request merged**